src/main/java/StringManipulation.java

```diff
@@ -1,33 +1,116 @@
+     * @return The string with the nth characters removed.
+     */
     @Override
     public String removeNthCharacter(int n, boolean maintainSpacing) {
-        return null;
-    }
+        if (n <= 0) {
+            throw new IllegalArgumentException();
+        } else if (n > string.length()) {
+            throw new IndexOutOfBoundsException();
+        }
 
-    @Override
-    public String[] getSubStrings(int startWord, int endWord) {
-        return null;
+        StringBuilder stringBuilder = new StringBuilder();
+
+        for (int i = 0; i < string.length(); i++) {
+            if ((i + 1) % n == 0) {
+                if (maintainSpacing) {
+                    stringBuilder.append(" ");
+                }
+            } else {
+                stringBuilder.append(string.charAt(i));
+            }
+        }
+
+        return stringBuilder.toString();
     }
 
+    /**
+     * Retrieves a substring of words from the specified startWord to the endWord.
+     * @param startWord The starting index of the substring.
+     * @param endWord The ending index of the substring.
+     * @return An array of substrings.
+     */
     @Override
-    public String restoreString(int[] indices) {
-        return null;
+    public String[] getSubStrings(int startWord, int endWord){
+        if (startWord <= 0 || endWord <= 0 || startWord > endWord) {
+            throw new IllegalArgumentException();
+        }
+
+        String[] preFilteredResult = string.split(" ");
+        String[] substrings = new String[endWord - startWord + 1];
+        int wordCount = 1;
+        int substringsIndex = 0;
+        
+        for (String result : preFilteredResult) {
+            if (!result.isEmpty()) {
+                if (wordCount >= startWord && wordCount <= endWord) {
+                    substrings[substringsIndex] = result;
+                    substringsIndex++;
+                }
+                wordCount++;
+            }
+        }
+
+        if (wordCount < endWord) {
+            throw new IndexOutOfBoundsException();
+        }
+
+        return substrings;
     }
 
+    /**
+     * Restores the characters from the given indices to form a string.
+     * @param indices The indices of the characters to be restored.
+     * @return The restored string.
+     */
+     @Override
+    public String restoreString(int[] indices){
+        if (indices.length != string.length()) {
+            throw new IllegalArgumentException();
+        }
+
+        String[] restoredStringArray = new String[indices.length];
+
+        for (int i = 0; i < indices.length; i++) {
+            int oldIndex = indices[i];
+            if (oldIndex < 0 || oldIndex >= string.length()) {
+                throw new IndexOutOfBoundsException();
+            }
+            restoredStringArray[i] = String.valueOf(string.charAt(oldIndex));
+        }
+
+        return String.join("", restoredStringArray);
```

Comment:
Your restoreString() basically has the same logic as mine and there is nothing wrong with your code:)
Just to share, instead of an array, a StringBuilder could be used as an alternative, which I used in my code. The same operation can be performed with append() for the StringBuilder, then return the string representation by using toString() on the StringBuilder.